Feature/login as a user

guerretta16 opened the pull request:
#3 
- RTK Api added for `Quote Hub`.
- Constants added for the URL and API_KEY `Quote Hub`.
- Types were added for `Quote Hub` response and `User` login form.
- Component Popup added to show the Login form.
- Component Login Form added.
   - Validates the username or email and password.
   - Execute 3 functions: `Get csrf token`, `Get access token` and `Get auth user info`.
- Refactor form Header. Popup component added there and if `access_token` exists, login button will hidde.
- Token slice was added with persist.
- Store was refactored to persist the info just for the `token`.

### RTK Endpoints

- Get csrf token. Used to protect the comunication with server.
- Get access token. (Login)
- Get auth user info.

guerretta16 committed:
feat(root): `redux-persist` library was added.

guerretta16 committed:
feat(store): `tokenSlice` and persist added.

- A token slice was added to persist the user token in localstorage.
- `store` file was modified to persist just the user token.
- Some barrel files were added too.

guerretta16 committed:
feat(layout): Popup and token validation in Header.

- Popup component added to login.
- If token exists, the login button will be hidden.

guerretta16 committed:
feat(utils): `Function` dir was added.

- Cookie file contains a function to extract a cookie passing the `cookie_name`.

guerretta16 committed:
feat(constants): `Quote Hub` constants added.

- Constants for the url and api key were added, and a tag too.

guerretta16 committed:
feat(types): `Quote Hub` and `User form` types were added

- Some types for the token, categories and user auth info were added in `QuoteHub`.
- Type for the `Login` was added into `User` file into entities.

guerretta16 committed:
feat(api): `Quote hub` api added.

- 3 endpoint were added into `quote hub` api.
   - Get access token.
   - Get csrf token (ensure connection).
   - Get auth user info.

- Some barrel files form app (redux slices), constants, api/rtk and utils.

guerretta16 committed:
feat(components): `LoginForm` and `ClickPopup` components added.

- ClickPopup is used to show the LoginForm as content.

- LoginForm has the next functions:
   - Validate the username or email and password.
   - Has the lazy hooks to get `csrf_token`, `access_token` and `auth user info`.
   - If csrf function fail, only shows a message 'A server error occurred!!!'.

f-esquivel commented on src/constants/quote_hub/quote_hub.constants.ts:
```diff
@@ -0,0 +1,4 @@
+export const QUOTE_HUB_API = {
+    API_KEY: import.meta.env.VITE_QUOTE_HUB_API_URL,
+    URL: 'http://localhost:8000/api/'
```
- [x] The `Quote Hub` API doesn't receive an `API_KEY` parameter.
- [x] The `URL`  constant should be a environment variable 😅

f-esquivel commented on src/api/rtk/quote_hub.api.ts:
```diff
@@ -0,0 +1,43 @@
+
+        })
+    })
+});
```
Refactor the `Quote Hub` RTK implementation.

More instructions on call.

f-esquivel commented on src/api/rtk/quote_hub.api.ts:
`quoteHub.api.ts`

Use camel case.

f-esquivel commented on src/app/features/token/tokenSlice.ts:
```diff
@@ -0,0 +1,25 @@
+    token: ''
+};
+
+export const tokenSlice = createSlice({
```
```suggestion
export const authSlice = createSlice({
```

Refactor all the files to use this name instead.

f-esquivel commented on src/components/LoginForm/LoginForm.tsx:
```diff
@@ -0,0 +1,84 @@
+export const LoginForm = () => {
+
+    const [getAccessToken, { isLoading: loadingAccessToken }] = useLazyGetAccessTokenQuery();
+    const [getCsrfToken, { isLoading: loadingCsrfToken }] = useLazyGetCsrfTokenQuery();
```
Call this query on a top level component like the main layout. All the API queries are going to need this variable.

f-esquivel commented on src/components/LoginForm/LoginForm.tsx:
```diff
@@ -0,0 +1,84 @@
+
+export const LoginForm = () => {
+
+    const [getAccessToken, { isLoading: loadingAccessToken }] = useLazyGetAccessTokenQuery();
```
Move to top level component.

f-esquivel commented on src/components/LoginForm/LoginForm.tsx:
```diff
@@ -0,0 +1,84 @@
+
+    const [getAccessToken, { isLoading: loadingAccessToken }] = useLazyGetAccessTokenQuery();
+    const [getCsrfToken, { isLoading: loadingCsrfToken }] = useLazyGetCsrfTokenQuery();
+    const [getAuthUserInfo] = useLazyGetAuthUserDataQuery();
```
Top level component

f-esquivel commented on src/constants/quote_hub/quote_hub.constants.ts:
`quoteHub.constants.ts` Use camel case.

f-esquivel commented on src/types/entities/QuoteHub.ts:
```diff
@@ -0,0 +1,17 @@
+}
+
+export type Categories = {
+    id: string
```
type id correctly with `CategoryType`

guerretta16 replied:
This id is for the DB info, for the relation between User and Category.
I just change the name, which should be a CategoryType.

f-esquivel commented on src/types/entities/index.ts:
```diff
@@ -1 +1,5 @@
-export * from './ApiNinjasQuote';
\ No newline at end of file
+export * from './ApiNinjasQuote';
+export * from './User';
+export * from './QuoteHub';
+export * from './User';
+export * from './QuoteHub';
```
This doesn't throw error? 🤔

f-esquivel commented on src/types/entities/User/User.ts:
```diff
@@ -0,0 +1,4 @@
+export interface UserLogin {
+    user_identifier: string;
+    password: string;
+}
```
Move this to a forms payloads types structure.

guerretta16 committed:
feat(api): `Endpoints` separated from rtk file.

- Authentication added into rtk `prepareHeaders`.
- Auth file added to inject endpoints.

guerretta16 committed:
refactor(slice): User state added for `AuthSlice`.

- File renamed, before `TokenSlice` now `AuthSlice`.
- Action added to save the user data into state.

guerretta16 committed:
refactor(*): Files removed constants, utils and types.

- Tags unused were removed.
- Function to get the CSRF Token were removed.
- Types of user form were removed.

guerretta16 committed:
refactor(api): `quoteHub` file were modified as base.

- Functions are going to be added as injections.

guerretta16 committed:
refactor(constants): `QuoteHub` constants were modified and URIs added.

- API_KEY constant was removed, 'cause isn't used.
- URL was refactored with env value.
- Auth uris were added as constants.

guerretta16 committed:
refactor(types): Category type was refactored and `Payload` file was added.

- Category type modified. This type is for the DB that returns the "id" and "name" which use `CategoryType` constant.

guerretta16 committed:
refactor(layouts): `MainLayout` and `HeaderNav` modified.

- Function to get the csrf token added into main layout.
- Function to get the user data added into main layout.

- In the header was modified the const to get the token from the state.

guerretta16 committed:
refactor(components): `LoginForm` was modified.

- Form data type was changed to `LoginPayload` type.
- Now just execute the function to get the access token.
- Message component from ant was removed.

guerretta16 committed:
refactor(store): `store` file was modified.

- Now, just the token is going to be persisted in localstorage.

guerretta16 committed:
fix(constants): tags constants for api ninja were deleted by accident.

- Tags were added and now it works.

f-esquivel reviewed (commented):
🤠👌🏾
